dev/test/transaction.ts

```diff
@@ -1042,6 +1073,177 @@ describe('failed transactions', () => {
   });
 });
 
+describe('successful transactions (with optimistic locking)', () => {
+  it('uses lastUpdateTime precondition for existing documents', () => {
+    const verify = {
+      verify: DOCUMENT_NAME,
+      currentDocument: {
+        updateTime: {
+          seconds: '3',
+          nanos: 4,
+        },
+      },
+    };
+
+    return runTransaction(
+      {optimisticLocking: true},
+      (transaction, docRef) => {
+        return transaction.get(docRef).then(doc => {
+          expect(doc.id).to.equal('documentId');
+        });
+      },
+      begin(),
+      getDocument(/* transaction= */ ''),
+      commit(undefined, [verify])
+    );
+  });
+
+  it('uses exists precondition for missing documents', () => {
+    const verify = {
+      verify: DOCUMENT_NAME,
+      currentDocument: {
+        exists: false,
+      },
+    };
+
+    return runTransaction(
+      {optimisticLocking: true},
+      (transaction, docRef) => {
+        return transaction.get(docRef).then(doc => {
+          expect(doc.exists).to.equal(false);
+        });
+      },
+      begin(),
+      getMissing(),
+      commit(undefined, [verify])
+    );
+  });
+
+  it('combines precondition with existing write', () => {
+    const set = {
+      update: {
+        fields: {},
+        name: DOCUMENT_NAME,
+      },
+      currentDocument: {
+        updateTime: {
+          seconds: '3',
+          nanos: 4,
+        },
+      },
+    };
+
+    return runTransaction(
+      {optimisticLocking: true},
+      (transaction, docRef) => {
+        return transaction.get(docRef).then(() => {
+          transaction.set(docRef, {});
+        });
+      },
+      begin(),
+      getDocument(/* transaction= */ ''),
+      commit(undefined, [set])
+    );
+  });
+
+  it('does not combines precondition if already set', () => {
+    const set = {
+      update: {
+        fields: {
+          a: {
+            mapValue: {
+              fields: {
+                b: {
+                  stringValue: 'c',
+                },
+              },
+            },
+          },
+        },
+        name: DOCUMENT_NAME,
+      },
+      updateMask: {
+        fieldPaths: ['a.b'],
+      },
+      currentDocument: {
+        updateTime: {
+          seconds: '1',
+          nanos: 2,
+        },
+      },
+    };
+    const verify = {
+      verify: DOCUMENT_NAME,
+      currentDocument: {
+        updateTime: {
+          seconds: '3',
+          nanos: 4,
+        },
+      },
+    };
+
+    return runTransaction(
+      {optimisticLocking: true},
+      (transaction, docRef) => {
+        return transaction.get(docRef).then(() => {
+          transaction.update(
+            docRef,
+            {'a.b': 'c'},
+            {lastUpdateTime: new Timestamp(1, 2)}
+          );
+        });
+      },
+      begin(),
+      getDocument(/* transaction= */ ''),
+      commit(undefined, [set, verify])
+    );
+  });
```

Comment:
optional (based on my above comments in transaction.ts): add test to check precondition if the same doc is read twice, add test to check that preconditions hold even if the document get() fails.

Reply:
Test added